npm/pkg/controlplane/controllers/v2/podcontroller.go

```diff
@@ -0,0 +1,633 @@
+			metrics.SendErrorLogAndMetric(util.PodID, "[podController processNextWorkItem] Error: failed to syncPod %s. Requeuing with err: %v", key, err)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		klog.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncPod compares the actual state with the desired, and attempts to converge the two.
+func (c *PodController) syncPod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the Pod resource with this namespace/name
+	pod, err := c.podLister.Pods(namespace).Get(name)
+
+	c.Lock()
+	defer c.Unlock()
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("pod %s not found, may be it is deleted", key)
+			// cleanUpDeletedPod will check if the pod exists in cache, if it does then proceeds with deletion
+			// if it does not exists, then event will be no-op
+			err = c.cleanUpDeletedPod(key)
+			if err != nil {
+				// need to retry this cleaning-up process
+				return fmt.Errorf("Error: %v when pod is not found\n", err)
+			}
+			return err
+		}
+
+		return err
+	}
+
+	// If newPodObj status is either corev1.PodSucceeded or corev1.PodFailed or DeletionTimestamp is set, start clean-up the lastly applied states.
+	if isCompletePod(pod) {
+		if err = c.cleanUpDeletedPod(key); err != nil {
+			return fmt.Errorf("Error: %v when when pod is in completed state.\n", err)
+		}
+		return nil
+	}
+
+	cachedNpmPod, npmPodExists := c.podMap[key]
+	if npmPodExists {
+		// if pod does not have different states against lastly applied states stored in cachedNpmPod,
+		// podController does not need to reconcile this update.
+		// in this updatePod event, newPod was updated with states which PodController does not need to reconcile.
+		if isInvalidPodUpdate(cachedNpmPod, pod) {
+			return nil
+		}
+	}
+
+	err = c.syncAddAndUpdatePod(pod)
+	if err != nil {
+		return fmt.Errorf("Failed to sync pod due to %v\n", err)
+	}
+
+	return nil
+}
+
+func (c *PodController) syncAddedPod(podObj *corev1.Pod) error {
+	klog.Infof("POD CREATING: [%s%s/%s/%s%+v%s]", string(podObj.GetUID()), podObj.Namespace,
+		podObj.Name, podObj.Spec.NodeName, podObj.Labels, podObj.Status.PodIP)
+
+	var err error
+	podNs := util.GetNSNameWithPrefix(podObj.Namespace)
+	podKey, _ := cache.MetaNamespaceKeyFunc(podObj)
+	// Add the pod ip information into namespace's ipset.
+	klog.Infof("Adding pod %s to ipset %s", podObj.Status.PodIP, podNs)
+	if err = c.dp.AddToSet([]string{podNs}, podObj.Status.PodIP, podKey); err != nil {
```

Comment:
Earlier addtoSet or any of theother functions would create ipset if it is missing, but the behavior now is DP expects a createSet call before any of these helper functions,